File: functions/middleware/checkFeature.js
const { getUserProfile } = require('../api/users');

/**
 * Middleware factory to check if user has access to a specific feature
 * @param {string} featureName - Name of the feature to check (e.g., 'analytics', 'postToX')
 * @returns {Function} Express middleware function
 */
const checkFeature = (featureName) => {
  return async (req, res, next) => {
    try {
      const { userId } = req;

      if (!userId) {
        return res.status(401).json({
          error: 'Unauthorized',
          message: 'User authentication required'
        });
      }

      // Get user profile
      const userProfile = await getUserProfile(userId);

      if (!userProfile) {
        return res.status(404).json({
          error: 'User Not Found',
          message: 'User profile not found'
        });
      }

      // Check if user has access to the feature
      const hasFeatureAccess = userProfile.features && userProfile.features[featureName] === true;

      if (!hasFeatureAccess) {
        return res.status(403).json({
          error: 'Upgrade to Pro/Business for this feature',
          feature: featureName,
          currentTier: userProfile.tier,
          message: `The '${featureName}' feature is not available on your current ${userProfile.tier} plan. Please upgrade to Pro or Business to access this feature.`
        });
      }

      // User has access, continue to next middleware/handler
      next();

    } catch (error) {
      console.error('Error checking feature access:', error);
      res.status(500).json({
        error: 'Internal Server Error',
        message: 'Failed to verify feature access'
      });
    }
  };
};

/**
 * Helper function to check multiple features at once
 * @param {string[]} featureNames - Array of feature names to check
 * @returns {Function} Express middleware function
 */
const checkMultipleFeatures = (featureNames) => {
  return async (req, res, next) => {
    try {
      const { userId } = req;

      if (!userId) {
        return res.status(401).json({
          error: 'Unauthorized',
          message: 'User authentication required'
        });
      }

      const userProfile = await getUserProfile(userId);

      if (!userProfile) {
        return res.status(404).json({
          error: 'User Not Found',
          message: 'User profile not found'
        });
      }

      // Check all required features
      const missingFeatures = featureNames.filter(feature => 
        !userProfile.features || userProfile.features[feature] !== true
      );

      if (missingFeatures.length > 0) {
        return res.status(403).json({
          error: 'Upgrade to Pro/Business for these features',
          features: missingFeatures,
          currentTier: userProfile.tier,
          message: `The following features are not available on your current ${userProfile.tier} plan: ${missingFeatures.join(', ')}. Please upgrade to Pro or Business.`
        });
      }

      next();

    } catch (error) {
      console.error('Error checking multiple feature access:', error);
      res.status(500).json({
        error: 'Internal Server Error',
        message: 'Failed to verify feature access'
      });
    }
  };
};

module.exports = {
  checkFeature,
  checkMultipleFeatures
};

File: quikthread-backend/middleware/check_feature.py
from fastapi import HTTPException, status, Depends
from middleware.auth import get_current_user
from services.user_service import UserService
from typing import Dict, Any
import logging

logger = logging.getLogger(__name__)

# Initialize user service
user_service = UserService()

def check_feature_access(feature_name: str):
    """
    Factory function to create feature access checker middleware
    
    Args:
        feature_name: Name of the feature to check (e.g., 'postToX', 'analytics')
        
    Returns:
        Dependency function that checks feature access
    """
    async def _check_feature(current_user: Dict[str, Any] = Depends(get_current_user)) -> str:
        """
        Check if user has access to a specific feature
        
        This middleware function:
        1. Gets the current authenticated user
        2. Retrieves their user profile from Firestore
        3. Checks if the feature is enabled for their tier
        4. Raises 403 error if feature not available
        5. Returns user_id if feature is available
        
        Args:
            current_user: Authenticated user from get_current_user dependency
            
        Returns:
            User ID if feature is available
            
        Raises:
            HTTPException: 403 if feature not available with upgrade message
        """
        try:
            user_id = current_user.get('uid')
            email = current_user.get('email')
            
            if not user_id:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="User ID not found in authentication token"
                )
            
            # Get user profile
            user_profile = await user_service.get_or_create_user(user_id, email or "unknown@example.com")
            
            # Check if feature is enabled
            features = user_profile.get('features', {})
            feature_enabled = features.get(feature_name, False)
            
            if not feature_enabled:
                tier = user_profile.get('tier', 'free')
                
                # Create helpful upgrade message based on feature
                feature_messages = {
                    'postToX': 'Direct posting to X (Twitter) is only available for Pro and Business tier users.',
                    'analytics': 'Analytics dashboard is only available for Business tier users.'
                }
                
                feature_message = feature_messages.get(
                    feature_name, 
                    f'The {feature_name} feature is not available on your current plan.'
                )
                
                logger.warning(f"Feature access denied for user {user_id}: {feature_name} (tier: {tier})")
                
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail={
                        "message": f"{feature_message} Please upgrade your plan to access this feature.",
                        "feature": feature_name,
                        "currentTier": tier,
                        "requiredTier": "pro" if feature_name == "postToX" else "business",
                        "upgradeUrl": "https://quikthread.com/pricing"
                    }
                )
            
            logger.info(f"Feature access granted for user {user_id}: {feature_name}")
            return user_id
            
        except HTTPException:
            # Re-raise HTTP exceptions
            raise
        except Exception as e:
            logger.error(f"Error checking feature access: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to verify feature access"
            )
    
    return _check_feature
